lte/gateway/c/oai/tasks/s1ap/s1ap_mme_handlers.c

```diff
@@ -1922,125 +1917,122 @@ static bool construct_s1ap_mme_full_reset_req(
+          "Reset Flag = %u\n",
+          assoc_id, reset);
+
       OAILOG_INFO(
-        LOG_S1AP,
-        "Moving eNB with association id %u to INIT state\n",
-        assoc_id);
+          LOG_S1AP, "Moving eNB with assoc_id %u to INIT state\n", assoc_id);
+      enb_association->s1_state = S1AP_INIT;
       update_mme_app_stats_connected_enb_sub();
     } else {
-      s1ap_remove_enb(state, enb_association);
-      update_mme_app_stats_connected_enb_sub();
       OAILOG_INFO(
-        LOG_S1AP,
-        "SCTP Shutdown request for association id %u. No Connected UEs.  = %u "
-        "\n",
-        assoc_id,
-        reset);
+          LOG_S1AP,
+          "SCTP Shutdown request for association id %u. No Connected UEs. "
+          "Reset Flag = %u\n",
+          assoc_id, reset);
+
       OAILOG_INFO(LOG_S1AP, "Removing eNB with association id %u \n", assoc_id);
+      s1ap_remove_enb(state, enb_association);
+      update_mme_app_stats_connected_enb_sub();
     }
     OAILOG_FUNC_RETURN(LOG_S1AP, RETURNok);
   }
 
+  OAILOG_INFO(
+      LOG_S1AP,
+      "SCTP disconnection request for association id %u, Reset Flag = "
+      "%u. Connected UEs = %d \n",
+      assoc_id, reset, enb_association->nb_ue_associated);
+
+  /*
+   * Send S1ap deregister indication to MME app in batches of UEs where
+   * UE count in each batch <= S1AP_ITTI_UE_PER_DEREGISTER_MESSAGE
+   */
   hashtable_uint64_ts_apply_callback_on_elements(
-    &enb_association->ue_id_coll,
-    s1ap_send_enb_deregistered_ind,
-    (void *) &arg,
-    (void **) &message_p);
+      &enb_association->ue_id_coll, s1ap_send_enb_deregistered_ind,
+      (void*) &arg, (void**) &message_p);
 
-  // The last batch of messages needs to be sent here
-  S1AP_ENB_DEREGISTERED_IND(message_p).nb_ue_to_deregister =
-    (uint8_t) arg.current_ue_index;
+  /*
+   * The last batch of messages needs to be sent here. Some UEs are left in the
+   * last batch to deregister if current_ue_index > 0
+   */
+  if ((uint8_t) arg.current_ue_index > 0) {
+    S1AP_ENB_DEREGISTERED_IND(message_p).nb_ue_to_deregister =
+        (uint8_t) arg.current_ue_index;
+
+    for (itrn = arg.current_ue_index;
+         itrn < S1AP_ITTI_UE_PER_DEREGISTER_MESSAGE; itrn++) {
+      S1AP_ENB_DEREGISTERED_IND(message_p).mme_ue_s1ap_id[itrn] = 0;
+      S1AP_ENB_DEREGISTERED_IND(message_p).enb_ue_s1ap_id[itrn] = 0;
+    }
+    S1AP_ENB_DEREGISTERED_IND(message_p).enb_id = enb_association->enb_id;
+    // Sending INVALID_IMSI64 because message is not specific to any UE/IMSI
+    message_p->ittiMsgHeader.imsi = INVALID_IMSI64;
 
-  for (i = arg.current_ue_index; i < S1AP_ITTI_UE_PER_DEREGISTER_MESSAGE; i++) {
-    S1AP_ENB_DEREGISTERED_IND(message_p).mme_ue_s1ap_id[arg.current_ue_index] =
-      0;
-    S1AP_ENB_DEREGISTERED_IND(message_p).enb_ue_s1ap_id[arg.current_ue_index] =
-      0;
+    itti_send_msg_to_task(TASK_MME_APP, INSTANCE_DEFAULT, message_p);
+    message_p = NULL;
```

Comment:
An easier and cleaner way to resolve this is by throwing "enb_association->ue_id_coll->num_elements" into arg_s1ap_send_enb_dereg_ind_t and let s1ap_send_enb_deregistered_ind handel the last batch

Reply:
Updated the PR. Now even the last batch of UEs  is handled by s1ap_send_enb_deregistered_ind function.